File: apps/light-desktop/src/api/client/macros.ts
import type {
	MacroCancelActionRequest,
	MacroExecutionSnapshot,
	MacroObjectAction,
	MacroObjectActionRequest,
	MacroRunActionRequest,
	MacroRunLineActionRequest,
	MacroRunLineUndoOutcome,
	MacroRuntimeSnapshot,
	MacroValidationRequest,
	FileInputClaimRequest,
	FileInputReleaseRequest,
	ShowObjectActionOutcome,
	MacroDefinition as WireMacroDefinition,
	MacroValidation as WireMacroValidation,
} from "../generated/light-wire";
import {
	type MacroExecution,
	type MacroRuntime,
	macroExecutionFromWire,
} from "../runtimeModels";
import { jsonRequest, type LiveClientTransport } from "./transport";

export interface MacroDefinition {
	id: string;
	number: number;
	name: string;
	source: string;
	presentation: { color?: string | null; icon?: string | null };
}

export interface MacroLineDiagnostic {
	line: number;
	status: "valid" | "invalid" | "interaction_required";
	message: string;
	tokens: Array<{
		start: number;
		end: number;
		kind:
			| "keyword"
			| "target"
			| "operator"
			| "address"
			| "number"
			| "timing"
			| "comment"
			| "text"
			| "definition";
		expansion?: string | null;
	}>;
}

export interface MacroSuggestion {
	label: string;
	insertText: string;
	detail: string;
	replaceStart: number;
	replaceEnd: number;
}

export interface MacroValidation {
	valid: boolean;
	diagnostics: MacroLineDiagnostic[];
	suggestions: MacroSuggestion[];
}

export class MacrosApiClient {
	constructor(private readonly transport: LiveClientTransport) {}

	validate(
		showId: string,
		source: string,
		cursor?: number,
	): Promise<MacroValidation> {
		const body: MacroValidationRequest = { source, cursor };
		return this.request<WireMacroValidation>(
			"/api/v2/macros/validate",
			showId,
			body,
		).then((validation) => ({
			valid: validation.valid,
			diagnostics: validation.diagnostics.map((diagnostic) => ({
				...diagnostic,
				tokens: diagnostic.tokens.map((token) => ({ ...token })),
			})),
			suggestions: (validation.suggestions ?? []).map((suggestion) => ({
				label: suggestion.label,
				insertText: suggestion.insert_text,
				detail: suggestion.detail,
				replaceStart: suggestion.replace_start,
				replaceEnd: suggestion.replace_end,
			})),
		}));
	}

	claimEditorInput(instanceId: string): Promise<void> {
		const body: FileInputClaimRequest = {
			request_id: crypto.randomUUID(),
			instance_id: instanceId,
			action: "macro_edit",
			origin: "editor",
		};
		return this.transport.request(
			"/api/v2/files/input-context/claim",
			jsonRequest("POST", body),
		);
	}

	releaseEditorInput(instanceId: string): Promise<void> {
		const body: FileInputReleaseRequest = {
			request_id: crypto.randomUUID(),
			instance_id: instanceId,
		};
		return this.transport.request(
			"/api/v2/files/input-context/release",
			jsonRequest("POST", body),
		);
	}

	mutate(
		showId: string,
		action: MacroObjectAction,
	): Promise<ShowObjectActionOutcome> {
		const body: MacroObjectActionRequest = {
			request_id: crypto.randomUUID(),
			action,
		};
		return this.request("/api/v2/macros/actions", showId, body);
	}

	create(showId: string, definition: MacroDefinition) {
		return this.mutate(showId, {
			type: "create",
			definition: definition as WireMacroDefinition,
		});
	}

	update(
		showId: string,
		macroId: string,
		expectedRevision: number,
		patch: Extract<MacroObjectAction, { type: "update" }>["patch"],
	) {
		return this.mutate(showId, {
			type: "update",
			macro_id: macroId,
			expected_revision: expectedRevision,
			patch,
		});
	}

	delete(showId: string, macroId: string, expectedRevision: number) {
		return this.mutate(showId, {
			type: "delete",
			macro_id: macroId,
			expected_revision: expectedRevision,
		});
	}

	copy(
		showId: string,
		sourceMacroId: string,
		expectedRevision: number,
		poolNumber: number,
	) {
		return this.mutate(showId, {
			type: "copy",
			source_macro_id: sourceMacroId,
			expected_revision: expectedRevision,
			pool_number: poolNumber,
		});
	}

	run(
		showId: string,
		macroId: string,
		request: MacroRunActionRequest = {},
	): Promise<MacroExecution> {
		void showId;
		return this.transport
			.sendAction({
				type: "macro",
				request: {
					type: "run",
					macro_id: macroId,
					source_revision: request.source_revision,
					trigger: request.trigger ?? { type: "web_socket" },
				},
			})
			.then((snapshot) => macroExecutionFromWire(snapshot as MacroExecutionSnapshot));
	}

	runLine(
		showId: string,
		macroId: string,
		request: MacroRunLineActionRequest,
	): Promise<MacroExecution> {
		void showId;
		return this.transport
			.sendAction({
				type: "macro",
				request: {
					type: "run_line",
					macro_id: macroId,
					source_revision: request.source_revision,
					line: request.line,
				},
			})
			.then((snapshot) => macroExecutionFromWire(snapshot as MacroExecutionSnapshot));
	}

	runtime(showId: string): Promise<MacroRuntime> {
		return this.transport
			.request<MacroRuntimeSnapshot>("/api/v2/macros/runtime", {
				headers: showHeaders(showId),
			})
			.then((snapshot) => ({
				desk_id: snapshot.desk_id,
				active: snapshot.active.map(macroExecutionFromWire),
				recent: snapshot.recent.map(macroExecutionFromWire),
			}));
	}

	execution(showId: string, executionId: string): Promise<MacroExecution> {
		return this.transport
			.request<MacroExecutionSnapshot>(
				`/api/v2/macros/executions/${encodeURIComponent(executionId)}`,
				{ headers: showHeaders(showId) },
			)
			.then(macroExecutionFromWire);
	}

	cancel(showId: string, executionId: string): Promise<MacroExecution> {
		const body: MacroCancelActionRequest = { execution_id: executionId };
		return this.request<MacroExecutionSnapshot>(
			"/api/v2/macros/executions/cancel",
			showId,
			body,
		).then(macroExecutionFromWire);
	}

	undoRunLine(
		showId: string,
		executionId: string,
	): Promise<MacroRunLineUndoOutcome> {
		return this.request(
			`/api/v2/macros/executions/${encodeURIComponent(executionId)}/undo-line`,
			showId,
			{},
		);
	}

	private request<T>(path: string, showId: string, body: unknown): Promise<T> {
		const request = jsonRequest("POST", body);
		return this.transport.request(path, {
			...request,
			headers: { ...request.headers, ...showHeaders(showId) },
		});
	}
}

function showHeaders(showId: string): HeadersInit {
	return { "x-tosk-show": showId };
}

File: crates/light/contracts/wire/src/v2/macros.rs
//! Show-owned command Macro object, validation, and one-shot runtime contracts.

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use ts_rs::TS;
use uuid::Uuid;

#[derive(Clone, Debug, Default, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroPresentation {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub color: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub icon: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroDefinition {
    pub id: Uuid,
    pub number: u16,
    pub name: String,
    pub source: String,
    #[serde(default)]
    pub presentation: MacroPresentation,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
#[serde(rename_all = "snake_case")]
pub enum MacroLineStatus {
    Valid,
    Invalid,
    InteractionRequired,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroToken {
    pub start: u32,
    pub end: u32,
    pub kind: MacroTokenKind,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub expansion: Option<String>,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
#[serde(rename_all = "snake_case")]
pub enum MacroTokenKind {
    Keyword,
    Target,
    Operator,
    Address,
    Number,
    Timing,
    Comment,
    Text,
    Definition,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroLineDiagnostic {
    pub line: u32,
    pub status: MacroLineStatus,
    pub message: String,
    pub tokens: Vec<MacroToken>,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroValidationRequest {
    pub source: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub cursor: Option<u32>,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroSuggestion {
    pub label: String,
    pub insert_text: String,
    pub detail: String,
    pub replace_start: u32,
    pub replace_end: u32,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroValidation {
    pub valid: bool,
    pub diagnostics: Vec<MacroLineDiagnostic>,
    #[serde(default)]
    pub suggestions: Vec<MacroSuggestion>,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroObjectActionRequest {
    pub request_id: String,
    pub action: MacroObjectAction,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum MacroObjectAction {
    Create {
        definition: MacroDefinition,
    },
    Update {
        macro_id: Uuid,
        #[ts(type = "number")]
        expected_revision: u64,
        patch: MacroPatch,
    },
    Copy {
        source_macro_id: Uuid,
        #[ts(type = "number")]
        expected_revision: u64,
        pool_number: u16,
    },
    Delete {
        macro_id: Uuid,
        #[ts(type = "number")]
        expected_revision: u64,
    },
}

#[derive(Clone, Debug, Default, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroPatch {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub number: Option<u16>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub source: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub presentation: Option<MacroPresentation>,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroObjectActionOutcome {
    pub request_id: String,
    pub replayed: bool,
    pub show_id: Uuid,
    #[ts(type = "number")]
    pub show_revision: u64,
    pub macro_id: Uuid,
    #[ts(type = "number")]
    pub object_revision: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub definition: Option<MacroDefinition>,
    pub deleted: bool,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroRunActionRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(as = "Option<f64>", optional = nullable)]
    pub source_revision: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub trigger: Option<MacroTrigger>,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroRunLineActionRequest {
    #[ts(type = "number")]
    pub source_revision: u64,
    pub line: u32,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum MacroLiveAction {
    Run {
        macro_id: Uuid,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        #[ts(as = "Option<f64>", optional = nullable)]
        source_revision: Option<u64>,
        trigger: MacroTrigger,
    },
    RunLine {
        macro_id: Uuid,
        #[ts(type = "number")]
        source_revision: u64,
        line: u32,
    },
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroCancelActionRequest {
    pub execution_id: Uuid,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroRunLineUndoOutcome {
    pub execution_id: Uuid,
    pub changed: bool,
    pub message: String,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum MacroTrigger {
    Pool,
    Editor,
    Playback { playback_number: u16 },
    CommandLine,
    Http,
    WebSocket,
    Osc,
    Hardware,
    Schedule,
    Timecode,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
#[serde(rename_all = "snake_case")]
pub enum MacroExecutionState {
    Queued,
    Validating,
    Running,
    Succeeded,
    Failed,
    Cancelled,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroExecutionSnapshot {
    pub execution_id: Uuid,
    pub macro_id: Uuid,
    pub macro_number: u16,
    pub macro_name: String,
    #[ts(type = "number")]
    pub source_revision: u64,
    pub desk_id: Uuid,
    pub session_id: Uuid,
    pub state: MacroExecutionState,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub line: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub statement: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub command: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub message: Option<String>,
    pub trigger: MacroTrigger,
    pub started_at: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub finished_at: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
pub struct MacroRuntimeSnapshot {
    pub desk_id: Uuid,
    pub active: Vec<MacroExecutionSnapshot>,
    pub recent: Vec<MacroExecutionSnapshot>,
}
